osquery/tables/system/windows/windows_security_products.cpp

```diff
@@ -0,0 +1,220 @@
+
+  // Loop over each product, querying the specific attributes.
+  for (LONG i = 0; i < ProductCount; i++) {
+    // Get the next security product
+    hr = PtrProductList->get_Item(i, &PtrProduct);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product item: " << hr;
+      goto exit;
+    }
+    wsc_entry tmp;
+    tmp.provider = provider;
+
+    // Get the product name
+    hr = PtrProduct->get_ProductName(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product name: " << hr;
+      goto exit;
+    }
+    tmp.product_name = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the product state
+    hr = PtrProduct->get_ProductState(&ProductState);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state: " << hr;
+      goto exit;
+    }
+    tmp.product_state = ProductState;
+
+    // Get the remediation path for the security product
+    hr = PtrProduct->get_RemediationPath(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get remediation path: " << hr;
+      goto exit;
+    }
+    tmp.remediation_path = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    // Get the signature status
+    hr = PtrProduct->get_SignatureStatus(&SignatureStatus);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get signature status: " << hr;
+      goto exit;
+    }
+    tmp.signature_status = SignatureStatus;
+
+    // Get the state timestamp
+    hr = PtrProduct->get_ProductStateTimestamp(&PtrVal);
+    if (FAILED(hr)) {
+      VLOG(1) << "Failed to get product state timestamp: " << hr;
+      goto exit;
+    }
+    tmp.product_state_timestamp = PtrVal;
+    SysFreeString(PtrVal);
+    PtrVal = nullptr;
+
+    PtrProduct->Release();
+    PtrProduct = nullptr;
+
+    out_list.push_back(tmp);
+  }
+
+exit:
+  if (nullptr != PtrVal) {
+    SysFreeString(PtrVal);
+  }
+  if (nullptr != PtrProductList) {
+    PtrProductList->Release();
+  }
+  if (nullptr != PtrProduct) {
+    PtrProduct->Release();
+  }
+  return hr;
+}
+
+void GetAllSecurityProducts(std::list<wsc_entry>& out_list) {
+  if (FAILED(GetSecurityProducts(WSC_SECURITY_PROVIDER_FIREWALL, out_list))) {
+    VLOG(1) << "Failed to get firewall products";
+  }
+  if (FAILED(GetSecurityProducts(WSC_SECURITY_PROVIDER_ANTIVIRUS, out_list))) {
+    VLOG(1) << "Failed to get antivirus products";
+  }
+  if (FAILED(
+          GetSecurityProducts(WSC_SECURITY_PROVIDER_ANTISPYWARE, out_list))) {
+    VLOG(1) << "Failed to get antispyware products";
+  }
+}
+
+namespace osquery {
+namespace tables {
+
+QueryData gen_wsp(QueryContext& context) {
+  QueryData results;
+  std::list<wsc_entry> products;
+  GetAllSecurityProducts(products);
+  // Use this to convert std::wstring into std::string
+  auto& str_converter = std::wstring_convert<std::codecvt_utf8<wchar_t>>();
+  for (const auto& product : products) {
+    Row r;
+
+    switch (product.provider) {
+    case WSC_SECURITY_PROVIDER_FIREWALL:
+      r["type"] = "Firewall";
```

Comment:
May I ask you to use translation maps instead of all this swich-case constructions?

Like:
```c++
const kWinSecurityProviders = std::unordered_map<int, std::string> {
  {WSC_SECURITY_PROVIDER_ANTIVIRUS, "Antivirus"},
  ...
};
...
r["type"] = tryTakeCopy(kWinSecurityProviders, product.provider).takeOr("Unknown");
```